File: solution2/emailXMLGen.pl
#!/usr/bin/perl

=head1 DESCRIPTION

Author:  Antonio A Eggermont
Date:    12/22/2012

emailXMLGen.pl - A simple XML generator script

To run this program:

    1. Make the location of perl in your system is correct in the script's shebang
    2. Change execute permissions to this script: 
           chmod a+x emailXMLGen.pl
    3. Then simply run ./emailXMLGen.pl

=cut


use strict;
use warnings;
use XML::Writer;
use IO;
use File::Basename qw( dirname );
use Cwd qw( abs_path );


my $emailList      = dirname(abs_path($0)) . "/emails.csv";             # Path to the emails lists data source
my $industryLookup = dirname(abs_path($0)) . "/industry_lookup.csv";    # Path to the industry associations data source
my $xmldoc         = dirname(abs_path($0)) . "/merged.xml";

sub lookupIndusAssoc{
    # Looks up for matching records in industry_lookup.csv file with
    # associated people email's address domains
    # ARGS: $email: Email address in processing 
    # RETURNS: Associated industry with person's email address DNS

    my $email = shift;
    my ($userame, $domain) = split(/@/, $email);

    open( FILEIN, "<:encoding(UTF-8)", $industryLookup)
        or die " cannot open file: $!";  

    while(<FILEIN>){
        next if ( $_ =~ m/Domain,Industry/ );
        my @indLookup = split(/,/, $_);

        if ($indLookup[0] =~ m/$domain/){
            $indLookup[1] =~ s/[\r\n]/ /g;
            return $indLookup[1];
	    }
    }

    return undef;
}


sub generateXMLDoc{
    # Generates an XML document based on data in %addressbook hash
    # ARGS:  \%addressbook: Reference address for %addressbook
    # RETURNS: None

    my $ref = shift;           # Referece address for %addressbook
    my %refAddr = %{$ref};

    my $doc = new IO::File(">$xmldoc");
    my $xmlObj = XML::Writer->new(OUTPUT => $doc, DATA_MODE => 1, DATA_INDENT => 4);
   
    $xmlObj->xmlDecl("UTF-8"); 
    $xmlObj->startTag("items");  

    foreach my $who ( keys %{refAddr} ) {
        $xmlObj->startTag("item");

        $xmlObj->startTag("name");
        $xmlObj->characters($who);
        $xmlObj->endTag("name");

        $xmlObj->startTag("email");
        $xmlObj->characters($refAddr{$who}{email});
        $xmlObj->endTag("email");

        $xmlObj->startTag("score");
        $xmlObj->characters($refAddr{$who}{score});  
        $xmlObj->endTag("score");

        if (defined $refAddr{$who}{industry}){
            $xmlObj->startTag("industry");
            $xmlObj->characters($refAddr{$who}{industry});  
            $xmlObj->endTag("industry");
        }

        $xmlObj->endTag("item");
    }
 
    $xmlObj->endTag("items");
    $xmlObj->end();   

}


#########################################################
#
#           Main program starts here  
#
#########################################################


my %addressbook;    # Holds a collection of records for people in emails.csv
my @record;         # Place holder to iterate over each record 

open(MYFILE, "<:encoding(UTF-8)", $emailList )
        or die "cannot open < $emailList : $!";

    while (<MYFILE>) {
        next if ( $_ =~ m/First Name/);
        $_ =~ s/\x0D//g;

        my @rec =  split(/,/, $_ );
        my $assoc = lookupIndusAssoc($rec[1]);
        $rec[2] =~ s/[\r\n]/ /g;
 
        $addressbook{$rec[0]} = {
            email    => $rec[1],
            score    => $rec[2],
            industry => $assoc
        };
    }

close(MYFILE)
        || warn "close failed: $!";



generateXMLDoc(\%addressbook);

print "XML file processed: ", $xmldoc, "\n";

File: solution1/emailXMLGen.py
#!/usr/bin/env python

"""
Author:  Antonio A Eggermont
Date:    12/22/2012

emailListGen.py - A simple XML generator script

To run this program:

    1. Change execute permissions to this script:
           chmod a+x emailListGen.py
    2. Then simply run ./emailListGen.py

"""

import os
import sys
import csv
import re
from xml.dom.minidom import Document


# Get fully qualified links for CSV source files
EMAILLIST = os.path.join(os.path.realpath(os.path.dirname(__file__)), 'emails.csv')
INDLOOKUP = os.path.join(os.path.realpath(os.path.dirname(__file__)), 'industry_lookup.csv')
XMLDOC    = os.path.join(os.path.realpath(os.path.dirname(__file__)), 'merged.xml')

addressbook = []  # Global container to hold person records

class Person:

    def __init__(self, name, email, score):

        self.name        = name
        self.attributes = dict([('email', email),
                                ('score', score),
                                ('association', None)])

    def addAssociation(self, association ):
        self.attributes['association'] = association

    def getName(self):
        return self.name

    def getEmail(self):
        return self.attributes['email']

    def getAttributes(self):
        return self.attributes

    def printMember(self):
        return [self.name ,
                self.attributes]


def load_data():
    """ Loads data from both email list and industry lookup csv source files  """

    try:
        for record in csv.reader(open(EMAILLIST), delimiter=','):
            if re.search(record[0], "First Name"): continue
            addressbook.append(Person( record[0] , record[1], record[2] ))

    except IOError as e:
        print "I/O error({0}): {1}".format(e.errno, e.strerror)
        raise


def process_data():
    """ Associates DNS portion of person's email address with industry look up sources
        and adds such association to the dictornary if present  """

    for member in addressbook:
        indAssoc = industry_lookup(member.getEmail())

        if indAssoc is not None:
            member.addAssociation(indAssoc)




def generate_xml_doc():
    """ Generates an XML tree document from records stored in addressbook dictionary  """

    doc = Document()
    root = doc.createElement('items')
    doc.appendChild(root)

    for member in addressbook:
        mainItem = doc.createElement('item')
        root.appendChild(mainItem)

        main = doc.createElement('name')
        mainItem.appendChild(main)
        text = doc.createTextNode(member.getName())
        main.appendChild(text)

        for itemKey, itemValue in member.getAttributes().iteritems():
            if str(itemKey) == "association" and str(itemValue) == "None": continue

            main = doc.createElement(str(itemKey))
            mainItem.appendChild(main)
            text = doc.createTextNode(str(itemValue))
            main.appendChild(text)


    # Lets do some clean up in the xml doc for pretty printing
    text_re = re.compile('>\n\s+([^<>\s].*?)\n\s+</', re.DOTALL)
    prettyXml = text_re.sub('>\g<1></', doc.toprettyxml(indent="    ",  encoding= 'utf-8'))

    try:
        with open(XMLDOC, "w") as fh:
            fh.write(prettyXml)
    except IOError as e:
        print "I/O error({0}): {1}".format(e.errno, e.strerror)
        raise
    finally:
        fh.close()



def industry_lookup(emailRecord):
    """ Does a look up in the industry association records to find out whether a match 
        is present with email's DNS portion  """

    parts = emailRecord.split(u'@')
    domain_part = parts[-1]

    try:
        for row in csv.reader(open(INDLOOKUP), delimiter=','):
            #if re.search(row[0], "Domain"): continue
            if re.match(domain_part, row[0]): return row[1]

    except IOError as e:
        print "I/O error({0}): {1}".format(e.errno, e.strerror)
        raise

    return None



if __name__ == "__main__":
    """   Program starts here  """

    load_data()
    process_data()
    generate_xml_doc()

    print "Processed XML File: %s" % (XMLDOC)
